Elgg 4 work

jeabakker opened the pull request:
feat(webservices): option to enable / disable API keys

chore(access): type/subtype is required for container permissions 
fixes #12684

fix(collections): corrected implementation of SeekableIterator
fixes #12558

feat(js): allow lightbox to load JS dependencies
fixes #12225

jeabakker committed:
feat(webservices): option to enable / disable API keys

jdalsem commented on mod/web_services/classes/ElggApiKey.php:
```diff
@@ -119,4 +119,36 @@ public function regenerateKeys() {
+	 *
+	 * @return bool
+	 */
+	public function isKeyActive() {
```
change function name to `hasActiveKeys` to keep plurar usage of the keys word intact

jdalsem commented on mod/web_services/classes/ElggApiKey.php:
```diff
@@ -119,4 +119,36 @@ public function regenerateKeys() {
+	 *
+	 * @return bool
+	 */
+	public function enableKey() {
```
change function name to `enableKeys` to keep plurar usage of the keys word intact

jdalsem commented on mod/web_services/classes/ElggApiKey.php:
```diff
@@ -119,4 +119,36 @@ public function regenerateKeys() {
+	 *
+	 * @return bool
+	 */
+	public function disableKey() {
```
change function name to `disableKeys` to keep plurar usage of the keys word intact

jdalsem commented on mod/web_services/languages/en.php:
```diff
@@ -73,6 +75,11 @@
 	'webservices:action:api_key:regenerate:success' => "The API keys have been regenerated",
 
+	'webservices:action:api_key:toggle_active:enable:success' => "The API key was successfully enabled",
+	'webservices:action:api_key:toggle_active:enable:error' => "An error occurred while enableing the API key",
```
typo: enabling

jdalsem commented on mod/web_services/languages/en.php:
```diff
@@ -73,6 +75,11 @@
 
+	'webservices:action:api_key:toggle_active:enable:success' => "The API key was successfully enabled",
+	'webservices:action:api_key:toggle_active:enable:error' => "An error occurred while enableing the API key",
+	'webservices:action:api_key:toggle_active:disable:success' => "The API key was successfullt disabled",
```
typo: successfully

jdalsem commented on mod/web_services/languages/en.php:
```diff
@@ -73,6 +75,11 @@
+	'webservices:action:api_key:toggle_active:enable:success' => "The API key was successfully enabled",
+	'webservices:action:api_key:toggle_active:enable:error' => "An error occurred while enableing the API key",
+	'webservices:action:api_key:toggle_active:disable:success' => "The API key was successfullt disabled",
+	'webservices:action:api_key:toggle_active:disable:error' => "An error occurred while disableing the API key",
```
typo: disabling

jdalsem commented on mod/web_services/views/default/object/api_key.php:
```diff
@@ -43,11 +43,27 @@
+
+if ($entity->isKeyActive()) {
+	$imprint[] = [
+		'icon_name' => 'check',
```
rewrite to use a ternary inline if/else function... saves a lot of lines of code.. can do for both icon_name and content

jeabakker replied:
don't agree

jdalsem commented on mod/web_services/views/default/object/api_key.php:
```diff
@@ -43,11 +43,27 @@
 	'byline' => false,
 	'access' => false,
 	'content' => $content,
+	'imprint' => $imprint,
```
if imprint was rewritten to ternary notation, you can add it directly here

jeabakker replied:
don't agree

jeabakker committed:
chore(access): type/subtype is required for container permissions

BREAKING CHANGE: when validating the container write permissions the
type and subtype of the content need to be provided.

fixes #12684

jeabakker committed:
fix(collections): corrected implementation of SeekableIterator

fixes #12558

jeabakker committed:
feat(js): allow lightbox to load JS dependencies

fixes #12225

jeabakker committed:
chore(db): disable query cache before fetching next chunk of ElggBatch

jeabakker committed:
chore(di): make subscription service a DI service

For future expansion this is easier

jdalsem commented on engine/classes/Elgg/Notifications/SubscriptionsService.php:
```diff
@@ -48,25 +56,27 @@ public function __construct(Database $db, array $methods = []) {
 	 * @return array
 	 */
-	public function getSubscriptions(NotificationEvent $event) {
+	public function getSubscriptions(NotificationEvent $event, array $methods = []) {
```
why default to an empty array?

jdalsem commented on engine/classes/Elgg/Notifications/SubscriptionsService.php:
```diff
@@ -90,18 +101,20 @@ public function getSubscriptions(NotificationEvent $event) {
-	public function getSubscriptionsForContainer($container_guid) {
-
-		$subscriptions = [];
+	public function getSubscriptionsForContainer(int $container_guid, array $methods = []) {
```
why default to empty array?

jdalsem commented on engine/classes/Elgg/Notifications/SubscriptionsService.php:
```diff
@@ -150,37 +164,35 @@ public function removeSubscription($userGuid, $method, $targetGuid) {
-
-		$container_guid = (int) $container_guid;
-
+	protected function getSubscriptionRecords(int $container_guid, array $methods = []) {
```
default empty array needed?

jdalsem commented on engine/classes/Elgg/UserCapabilities.php:
```diff
@@ -244,13 +244,13 @@ public function canEditAnnotation(ElggEntity $entity, $user_guid = 0, ElggAnnota
 	 * @return bool
 	 */
-	public function canWriteToContainer(ElggEntity $entity, $user_guid = 0, $type = 'all', $subtype = 'all') {
+	public function canWriteToContainer(ElggEntity $entity, string $type, string $subtype, int $user_guid = 0) {
```
do we need to document the order change? we documented other changes from this class in the upgrade notes

jeabakker replied:
removed other docs
